python3/httplib2/__init__.py

```diff
@@ -836,27 +914,92 @@ class HTTPSConnectionWithTimeout(http.client.HTTPSConnection):
-                check_hostname=disable_ssl_certificate_validation ^ True)
+                self.proxy_info = proxy_info('https')
+
+        context = ssl.SSLContext(self.ssl_version)
+        context.verify_mode = ssl.CERT_NONE if disable_ssl_certificate_validation else ssl.CERT_REQUIRED
+        context.check_hostname = not disable_ssl_certificate_validation
+
+        self.ca_certs = ca_certs if ca_certs else CA_CERTS
+        context.load_verify_locations(self.ca_certs)
+
+        if cert_file:
+            context.load_cert_chain(cert_file, key_file)
+
+        super(HTTPSConnectionWithTimeout, self).__init__(host, port=port, key_file=key_file, cert_file=cert_file,
+                                                         timeout=timeout, context=context)
+
+    def connect(self):
+        """Connect to a host on a given (SSL) port."""
+        if self.proxy_info and self.proxy_info.isgood():
+            use_proxy = True
+            proxy_type, proxy_host, proxy_port, proxy_rdns, proxy_user, proxy_pass, proxy_headers = self.proxy_info.astuple()
+
+            host = proxy_host
+            port = proxy_port
+        else:
+            use_proxy = False
+
+            host = self.host
+            port = self.port
+            proxy_type = None
+            proxy_headers = None
+
+        address_info = socket.getaddrinfo(host, port, 0, socket.SOCK_STREAM)
+        for family, socktype, proto, canonname, sockaddr in address_info:
+            try:
+                if use_proxy:
+                    sock = socks.socksocket(family, socktype, proto)
+
+                    sock.setproxy(proxy_type, proxy_host, proxy_port, proxy_rdns, proxy_user, proxy_pass)
+                else:
+                    sock = socket.socket(family, socktype, proto)
+                    sock.setsockopt(socket.IPPROTO_TCP, socket.TCP_NODELAY, 1)
+                if has_timeout(self.timeout):
+                    sock.settimeout(self.timeout)
+                sock.connect((self.host, self.port))
+                self.sock = ssl_wrap_socket(
+                    sock, self.key_file, self.cert_file,
+                    self.disable_ssl_certificate_validation, self.ca_certs,
+                    self.ssl_version, self.host)
+                if self.debuglevel > 0:
+                    print("connect: ({0}, {1})".format(self.host, self.port))
+                    if use_proxy:
+                        print("proxy: {0}".format(str(
+                            (proxy_host, proxy_port, proxy_rdns, proxy_user, proxy_pass, proxy_headers))))
+            except (ssl.SSLError, ssl.CertificateError) as e:
+                if sock:
+                    sock.close()
+                if self.sock:
+                    self.sock.close()
+                self.sock = None
+                raise
+            except (socket.timeout, socket.gaierror):
+                raise
+            except socket.error as error_msg:
+                self.error_msg = error_msg
```

Comment:
- (minor, naming) it is not error message, `except` statement binds exception object. Error message is, sometimes, `ex.msg` attribute of exception object, or more generally, it's `str(ex)`. But we don't need error message here. We need actual exception object.
- (major) assigning to instance leaks exception object (and associated traceback) after `connect()` call
- (major) `raise socket.error(self.error_msg)` is creating another exception object wrapping previous one, so caller will not get `.errno` and maybe other useful attributes of original exception object
- (major) `self.error_msg` leaks code flow, given how complicated code is here, it's possible that `raise socket.error(self.error_msg)` will happen with outdated, irrelevant object

Solution: define error object in scope of `connect()` function and raise it without wrapping into another level of `socket.error`
Something like this:

```python
def connect(self):
  socket_err = None
  # ...
    try:
    # ...
    except socket.error as e:
      socket_err = e
      # if debug...
  if not self.sock:
    raise socket_err
```

I'm sorry for headaches for code that you didn't write just copied from other file, but it's really better to fix it earlier than later. We have outstanding issue with this except/socket.error in py2 code tree.